### PR TITLE
Exposure functions, Survival and Resistance

### PR DESCRIPTION
**Pull request by virgile-baudrot**
# Description

[Describe what this PR adds to the package]

## Changes (choose one only, with x or click after creating PR):

- [x] New formulation/rule
- [ ] Documentation
- [ ] Bug fix for existing rules (non-breaking syntax)
- [ ] Major change to existing rules (breaking syntax, needs a breaking version bump)

# Checklist:

- [ ] My code follows the [BlueStyle](https://github.com/invenia/BlueStyle) style guide
- [x] My code is commented with explanations of anything hard to understand
- [x] There are no unnecessary or unrelated code changes included in this PR
- [x] There are no depencecies added: if so, please consider if they are necessary and explain why they are required

### For bugfixes

- [ ] My PR addresses a single bug/connected group of bugs
- [ ] I have included a test that will ensure the bug does not recur

### For new `Rule`s

- [ ] My PR is limited to a clear theme/cenceptual unit. (Otherwise make multiple PRs that can be discussed separately)
- [x] New rules have tests in a file of the same name as the rule file, e.g. `/test/newrulename.jl`
- [ ] I have cited the source of the new formulation/s, if applicable
- [ ] I have added a documentation string to formulation struct/s, and added an entry in the docs

### CODE REVIEW

**Comment by virgile-baudrot** [comment]
It's just a Draft pull request, I've to check all the PR points are ok to be ready for review

**Comment by virgile-baudrot** [comment]
well, I'm going to revise my PR...

File: src/Dispersal.jl [hunk 1]
```diff
@@ -34,6 +34,14 @@ export HumanDispersal, BatchGroups, HeirarchicalGroups
 
 export ExponentialGrowth, LogisticGrowth, ThresholdGrowth
 
+export LogisticSurvival
+
+export Threshold_Exposure, Rotation_Exposure
```
Comment (rafaqz) [inline src/Dispersal.jl:39]: No underscores in type names - so `RotationExposure`

File: src/exposure.jl [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    degradationRate::DR
+    "Timestep used in formulation"
+    timestep::TS
+    "The fractional number of rule timesteps in the current simulation timestep"
+    nsteps::S
+end
+Threshold_Exposure{R,W}(;
+    crop=CROP_OCCUPANCY,
+    pulseLevel=PULSE_LEVEL,
+    popThreshold=POPULATION_THRESHOLD,
```
Comment (rafaqz) [inline src/exposure.jl:43]: snake case for fields: `pop_threshold`, or better `popthreshold`

File: src/exposure.jl [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    # return pesticide only
+    if population < rule.popThreshold
+        @fastmath pesticide * exp(- rule.degradationRate * rule.nsteps) 
+    else 
+        @fastmath pesticide * exp(- rule.degradationRate * rule.nsteps) + rule.pulseLevel
+    end
+end
+
+"""
+Exposure by several treatments in rotation depending on a threshold layer (population size) set with `popThreshold`.
```
Comment (rafaqz) [inline src/exposure.jl:63]: Add a signature for the constructor so people can see how to build this struct

File: src/exposure.jl [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+"""
+Extends CellRule for rules of exposure dynamics
+
+For best performance these should be chained with other
+CellRule or following an NeighborhoodRule.
+"""
+
+abstract type Exposure{R,W} <: CellRule{R,W} end
+
+"""
```
Comment (rafaqz) [inline src/exposure.jl:17]: Add a signature for the constructor in the doc string, something like:

```julia
Threshold_Exposure{R,W}(;
    crop,  pulseLevel,  popThreshold,  degradationRate,  timestep,
)
```
See the other files for examples.

File: src/exposure.jl [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+Exposure by treatment depending on a threshold layer (population size) set with `popThreshold`.
+Treatment is applied only on `crop` fields.
+The amount of pesticide `pulseLevel` is applied once this threshold is passed.
+At every time step the pesticide is degradated at rate `degradationRate`
+
+Read: Tuple(pesticide, population)
+Written: Tuple(pesticide)
+"""
+struct Threshold_Exposure{R,W,CP,PL,PT,DR,TS,S} <: Exposure{R,W}
+    "key layer for occurence of plot"
```
Comment (rafaqz) [inline src/exposure.jl:27]: Can remove these doc strings and instead write them in the main doc string under an `## Arguments` header.

```julia
- `popthreshold`: some description
```

File: src/exposure.jl [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+end
+
+struct RotationStruct{RS,TS}
+    rotationStep::RS
+    timeStep::TS
+end
+
+# All numerical operation are on the first element
+Base.:*(rs::RotationStruct, x::Number) = RotationStruct(x * rs.rotationStep, rs.timeStep) 
+Base.:*(x::Number, rs::RotationStruct) = RotationStruct(x * rs.rotationStep, rs.timeStep)
```
Comment (rafaqz) [inline src/exposure.jl:140]: Will also need one method for division by a scalar.

File: src/exposure.jl [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+Base.:+(rs::RotationStruct, x::Number) = RotationStruct(x + rs.rotationStep, rs.timeStep) 
+Base.:+(x::Number, rs::RotationStruct) = RotationStruct(x + rs.rotationStep, rs.timeStep)
+Base.:+(rs1::RotationStruct, rs2::RotationStruct) = RotationStruct(rs1.rotationStep + rs2.rotationStep, rs1.timeStep + rs2.timeStep)
+Base.:-(rs::RotationStruct, x::Number) = RotationStruct(rs.rotationStep - x, rs.timeStep)
```
Comment (rafaqz) [inline src/exposure.jl:144]: Probably don't need to add or subtract scalars?

File: src/exposure.jl [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+Base.:+(rs::RotationStruct, x::Number) = RotationStruct(x + rs.rotationStep, rs.timeStep) 
+Base.:+(x::Number, rs::RotationStruct) = RotationStruct(x + rs.rotationStep, rs.timeStep)
+Base.:+(rs1::RotationStruct, rs2::RotationStruct) = RotationStruct(rs1.rotationStep + rs2.rotationStep, rs1.timeStep + rs2.timeStep)
+Base.:-(rs::RotationStruct, x::Number) = RotationStruct(rs.rotationStep - x, rs.timeStep) 
+Base.:-(x::Number, rs::RotationStruct) = RotationStruct(x - rs.rotationStep, rs.timeStep)
+Base.:-(rs1::RotationStruct, rs2::RotationStruct) = RotationStruct(rs1.rotationStep - rs2.rotationStep, rs1.timeStep - rs2.timeStep)
+
+Base.mod(rs::RotationStruct, x::Number) = mod(rs.rotationStep, x)
+
+isEqualStep(rs::RotationStruct, y::Number) = rs.timeStep == y ? true : false
```
Comment (rafaqz) [inline src/exposure.jl:150]: `isequalstep` or `isequal_step` or something if you really need an underscore. No capitals in function names

File: src/resistance.jl [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+const DEVIATION_PHENOTYPE = Param(0.0; bounds=(0.0, 100.0))
+const DOMINANCE_DEGREE = Param(0.0; bounds=(0.0, 100.0))
+
+"""
+Extends Resistance for using heterogenous data.
+"""
+abstract type Resistance{R,W} <: CellRule{R,W} end
+
+"""
+Lande Equation application
```
Comment (rafaqz) [inline src/resistance.jl:11]: Constructor signature at the top of the docs, more explanation of the equation (citation if possible), and list of the arguments as above.

File: src/survival.jl [hunk 10]
```diff
@@ -0,0 +1,38 @@
+const LC50_PARAM = Param(0.0; bounds=(0.0, 1e9))
+const HILLCOEFFICIENT_PARAM = Param(0.1; bounds=(0.0, 100))
+
+"""
+Extends CellRule for rules of survival effect
```
Comment (rafaqz) [inline src/survival.jl:5]: Can remove the performance hint its not always true now! (the general performance got too good so chaining rules might slow it down by increasing the page size of the function)

File: src/survival.jl [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+For best performance these should be chained with other
+CellRule or following an NeighborhoodRule.
+"""
+abstract type Survival{R,W} <: CellRule{R,W} end
+
+"""
+Logistic Survival.
```
Comment (rafaqz) [inline src/survival.jl:12]: Add the constructor signature and the list of arguments

File: test/resistance.jl [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+             0.0 0.0 0.0]
+PopAlleleFreq = population .* pFreqinit # all are SS
+
+meanPhenotype = 30.0
+deviationPhenotype = 5.0
+dominanceDegree = 10.0
+PopPhenotype = population .* (meanPhenotype .+ pFreqinit.^2 .* deviationPhenotype + 2 .* pFreqinit .*(1 .-pFreqinit) .* dominanceDegree - (1 .-pFreqinit).^2 .* deviationPhenotype)
+
+init = (
+    pesticide =  [0.0 0.0 0.0;
```
Comment (rafaqz) [inline test/resistance.jl:23]: Make the columns line up

File: test/resistance.jl [hunk 13]
```diff
@@ -0,0 +1,49 @@
+using Dispersal
+using DynamicGrids, Unitful, Dates
+using DimensionalData
+using Test
+
+@testset "Test resistance Lande equations" begin
+
+population = [  100000.0 100.0 0.0;
+                0.0 1000.0 0.0;
```
Comment (rafaqz) [inline test/resistance.jl:9]: Make the matrix columns line up, or use a single line.

File: test/resistance.jl [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+pFreqinit = [1.0 0.0 0.0;
+             0.0 0.0 0.0;
+             0.0 0.0 0.0]
+PopAlleleFreq = population .* pFreqinit # all are SS
+
+meanPhenotype = 30.0
+deviationPhenotype = 5.0
+dominanceDegree = 10.0
+PopPhenotype = population .* (meanPhenotype .+ pFreqinit.^2 .* deviationPhenotype + 2 .* pFreqinit .*(1 .-pFreqinit) .* dominanceDegree - (1 .-pFreqinit).^2 .* deviationPhenotype)
```
Comment (rafaqz) [inline test/resistance.jl:20]: Line is too lone

File: test/exposure.jl [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    ruleTreatment2 = Rotation_Exposure{Tuple{:pesticide2,:population,:rotation}, Tuple{:pesticide2,:rotation}}(
+        crop=Aux(:crop), rotationSize=3, rotationNumber= 2, pulseLevel=20.0, popThreshold = 1000.0,degradationRate=0.5
+    )
+    ruleTreatment3 = Rotation_Exposure{Tuple{:pesticide3,:population,:rotation}, Tuple{:pesticide3,:rotation}}(
+        crop=Aux(:crop), rotationSize=3, rotationNumber= 3, pulseLevel=20.0, popThreshold = 1000.0,degradationRate=0.5
+    )
+    rule = Ruleset(Chain(ruleTreatment1, ruleTreatment2,ruleTreatment3))
+    sim!(output, rule)
+
+    CheckReturn(i) = [output[i].pesticide1[1] , output[i].pesticide2[1], output[i].pesticide3[1]]
```
Comment (rafaqz) [inline test/exposure.jl:115]: functions are lower case/snake case: `checkreturn(i)`. This looks like a type constructor to me.

File: test/exposure.jl [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+        population = [100000.0 100.0 0.0; 0.0 1000.0 0.0; 0.0 0.0 100000.0])
+
+end   
+
+##################
+# GENERIC ROTATION
+##################
+
+@testset "rotation structure" begin
+    RS(x,y) = Dispersal.RotationStruct(x,y)
```
Comment (rafaqz) [inline test/exposure.jl:64]: This uppercase method is fine, as its actually a constructor and it suggests that.

File: test/exposure.jl [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    @test output[Ti(3)] == (
+        pesticide = [0.0 0.0 0.0; 0.0 32.13061319425267 0.0; 0.0 0.0 32.13061319425267],
+        population = [100000.0 100.0 0.0; 0.0 1000.0 0.0; 0.0 0.0 100000.0])
+
+end   
+
+##################
+# GENERIC ROTATION
+##################
+
```
Comment (rafaqz) [inline test/exposure.jl:62]: Add some comments to this, its hard to understand what this is for. You can also make the testset descriptions more descriptive and make them nested, if that helps.

File: test/exposure.jl [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+   
+    @test mod(RS(1,2),2) == mod(1,2)
+    @test mod(RS(2,2),2) == mod(2,2)
+    @test mod(RS(2,3),2) == mod(2,2)
+    @test mod(RS(3,2),2) == mod(3,2)
+
+    @test RS(3,2) + 1 == RS(4,2)
+    @test RS(3,2) + 8.3 == RS(11.3,2)
+
+    US(x,y) = Dispersal.updateStep(x, y)
```
Comment (rafaqz) [inline test/exposure.jl:74]: this isn't a constructor, it should be lower case (both US and updateStep)

File: test/exposure.jl [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+
+end   
+
+##################
+# GENERIC ROTATION
+##################
+
+@testset "rotation structure" begin
+    RS(x,y) = Dispersal.RotationStruct(x,y)
+
```
Comment (rafaqz) [inline test/exposure.jl:65]: Put this in a testset with a descriptive name of what is being tested. The tests look good, but its not immediately clear what you are testing

```julia
@testset "can take modulo of RotationStruct" begin # Or something, I'm not sure what it does!
    @test mod(RS(1,2),2) == mod(1,2)
    @test mod(RS(2,2),2) == mod(2,2)
    @test mod(RS(2,3),2) == mod(2,2)
    @test mod(RS(3,2),2) == mod(3,2)
end
```

File: test/growth.jl [hunk 20]
```diff
@@ -132,6 +134,33 @@ end
 
 end
 
+@testset "logistic growth with rate from a growth 3D map" begin
+
+    init =  [1.0 1.0 1.0;
+        1.0 1.0 1.0;
```
Comment (rafaqz) [inline test/growth.jl:140]: line up the matrix or make it one line

File: test/growth.jl [hunk 21]
```diff
@@ -132,6 +134,33 @@ end
...
+        1.0 1.0 1.0]
+   
+    suit =  reshape(
+        repeat(
+        [1.0 1.0 2.0;
+        2.0 2.0 0.5;
```
Comment (rafaqz) [inline test/growth.jl:146]: line up the matrix or make it one line

File: test/resistance.jl [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+output = ArrayOutput(init; tspan=1:3,  aux=(crop=crop,))
+ruleTreatment = Threshold_Exposure{Tuple{:pesticide,:population}, :pesticide}(crop=Aux(:crop), pulseLevel=20, popThreshold = 1000)
+ruleGrowth = LogisticGrowth{:population,:population}(rate=5)
+ruleSurvival = LogisticSurvival{Tuple{:population,:popPhenotype,:pesticide}, :population}(hillcoefficient=hillcoefficient)
+ruleResistance = Lande_Resistance{Tuple{:popAlleleFreq,:popPhenotype,:population,:pesticide}, Tuple{:popAlleleFreq,:popPhenotype}}(
+    hillcoefficient=hillcoefficient, deviationPhenotype=deviationPhenotype, dominanceDegree=dominanceDegree)
+
+rule = Ruleset(ruleGrowth, ruleTreatment, ruleSurvival, ruleResistance)
+sim!(output, rule)
+output[Ti(1)]
```
Comment (rafaqz) [inline test/resistance.jl:45]: Should these be tests?

File: test/resistance.jl [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+@testset "Test resistance Lande equations" begin
+
+population = [  100000.0 100.0 0.0;
+                0.0 1000.0 0.0;
+                0.0 0.0 100000.0]
+
+pFreqinit = [1.0 0.0 0.0;
+             0.0 0.0 0.0;
+             0.0 0.0 0.0]
+PopAlleleFreq = population .* pFreqinit # all are SS
```
Comment (rafaqz) [inline test/resistance.jl:15]: variables are lowercase or snake_case

File: test/exposure.jl [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+        pesticide1 = zeros(nr,nc),
+        pesticide2 = zeros(nr,nc),
+        pesticide3 = zeros(nr,nc),
+        rotation = Rotation_InitGrid(1, (nr,nc)),
+        population = ones(nr,nc).*10^4, )
+
+    crop = ones(nr,nc)
+       
+    output = ArrayOutput(init; tspan=1:7,  aux=(crop=crop,))
+    ruleTreatment1 = Rotation_Exposure{Tuple{:pesticide1,:population,:rotation}, Tuple{:pesticide1,:rotation}}(
```
Comment (rafaqz) [inline test/exposure.jl:103]: variables are lowercase or snake_case: `rule_treatment1`

File: test/exposure.jl [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+    sim!(output, rule)
+    @test output[Ti(1)] == (
+        pesticide = [0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0],
+        population = [100000.0 100.0 0.0; 0.0 1000.0 0.0; 0.0 0.0 100000.0])
+    @test output[Ti(2)] == (
+        pesticide = [0.0 0.0 0.0; 0.0 20.0 0.0; 0.0 0.0 20.0],
+        population = [100000.0 100.0 0.0; 0.0 1000.0 0.0; 0.0 0.0 100000.0])
+    @test output[Ti(3)] == (
+        pesticide = [0.0 0.0 0.0; 0.0 32.13061319425267 0.0; 0.0 0.0 32.13061319425267],
+        population = [100000.0 100.0 0.0; 0.0 1000.0 0.0; 0.0 0.0 100000.0])
```
Comment (rafaqz) [inline test/exposure.jl:55]: NamedTuple syntax:
```julia
(
        pesticide=[0.0 0.0 0.0; 0.0 32.13061319425267 0.0; 0.0 0.0 32.13061319425267],
        population=[100000.0 100.0 0.0; 0.0 1000.0 0.0; 0.0 0.0 100000.0]
)
```

No spaces in `x=y`, to distinguish it from variable assignment. And the bracket outside the list, although I don't really mind either way.

File: test/exposure.jl [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+        population = [100000.0 100.0 0.0;
+                    0.0 1000.0 0.0;
+                    0.0 0.0 100000.0], )
+
+    crop = [0.0 1.0 1.0; 
+    0.0 10.0 0.0;
+    0.0 0.0 10.0]
+        
+    output = ArrayOutput(init; tspan=1:3,  aux=(crop=crop,))
+    ruleTreatment = Threshold_Exposure{Tuple{:pesticide,:population}, :pesticide}(crop=Aux(:crop), pulseLevel=20, popThreshold = 1000)
```
Comment (rafaqz) [inline test/exposure.jl:21]: `rule_treatment`

File: test/exposure.jl [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+using DimensionalData
+using Test
+
+@testset "Test exposure population" begin
+
+    init = (
+        pesticide =  [0.0 0.0 0.0; 
+                    0.0 0.0 0.0;
+                    0.0 0.0 0.0],
+        population = [100000.0 100.0 0.0;
```
Comment (rafaqz) [inline test/exposure.jl:12]: line up the matrices or make them one line.

File: test/exposure.jl [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+        population = [100000.0 100.0 0.0; 0.0 1000.0 0.0; 0.0 0.0 100000.0])
+    @test output[Ti(2)] == (
+        pesticide = [0.0 0.0 0.0; 0.0 20.0 0.0; 0.0 0.0 20.0],
+        population = [100000.0 100.0 0.0; 0.0 1000.0 0.0; 0.0 0.0 100000.0])
+    @test output[Ti(3)] == (
+        pesticide = [0.0 0.0 0.0; 0.0 32.13061319425267 0.0; 0.0 0.0 32.13061319425267],
+        population = [100000.0 100.0 0.0; 0.0 1000.0 0.0; 0.0 0.0 100000.0])
+
+    # If cropping change allong time
+    crop = reshape( # TODO: something that change with time
```
Comment (rafaqz) [inline test/exposure.jl:35]: If this TODO is important, do it now ;) 

No one will do it later!

File: test/exposure.jl [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+        RS(1,DateTime(2020,2,1)) RS(1,DateTime(2020,2,1))
+    ]
+end
+
+
+@testset "Test rotation in exposure" begin
+    nr, nc = 1, 1
+    init = (
+        pesticide1 = zeros(nr,nc),
+        pesticide2 = zeros(nr,nc),
```
Comment (rafaqz) [inline test/exposure.jl:95]: always a space after commas

File: src/exposure.jl [hunk 30]
```diff
@@ -0,0 +1,161 @@
...
+) where {R,W} = Threshold_Exposure{R,W}(crop, pulseLevel, popThreshold, degradationRate, timestep, nsteps)
+
+precalcrule(rule::Threshold_Exposure, data) = precalc_timestep(rule, data)
+
+@inline function applyrule(data, rule::Threshold_Exposure, (pesticide, population), I)
+    crop = get(data, rule.crop, I...) 
+    crop > zero(crop) || return zero(pesticide) 
+    # return pesticide only
+    if population < rule.popThreshold
+        @fastmath pesticide * exp(- rule.degradationRate * rule.nsteps)
```
Comment (rafaqz) [inline src/exposure.jl:56]: no space after minus sign here

File: src/Dispersal.jl [hunk 31]
```diff
@@ -51,5 +59,8 @@ include("growth.jl")
 include("human.jl")
 include("jump.jl")
 include("allee.jl")
+include("exposure.jl")
```
Comment (rafaqz) [inline src/Dispersal.jl:62]: probably one file per PR is easier here, this is much too long

**Review by rafaqz — CHANGES_REQUESTED** [review]
Ok reviewed this, code and tests look nice, but not BlueStyle and its very long!

Can you do exposure.jl, survival.jl, and resistance.jl separately in new PRs? incorporating these comments and applying BlueStyle to everything. 

I'm mostly repeating myself throughout and there are too many things to take in and discuss. I think I misunderstood this and you were going to change that anyway, but hopefully this is helpful too :))

Really we should be just talking about the formulation and whether the tests are good enough.

**Comment by virgile-baudrot** [comment]
Fine, I close this one and reopen new oneS with your suggestions 👍

**Pull request closed without merging**